### PR TITLE
feat: Add channel priority to solve task and expose to python solve

### PR DESCRIPTION
**Pull request by BenjaminLowry**
There are use cases where disabling strict channel priority is important and so should have functionality in rattler such that it can be disabled. So I added it as a field of the SolveTask and exposed it as a new parameter (that existed before but was deleted) in py-rattler.

### CODE REVIEW

File: crates/rattler_solve/src/libsolv_c/mod.rs
Comment (BenjaminLowry) [inline crates/rattler_solve/src/libsolv_c/mod.rs:99]: This part I'm not so sure what we'd want the behavior to be, it seems to me that this libsolv code doesn't enforce strict channel priority? If so, we can either ignore the option all together to throw if strict priority is true since that's "not supported"?

**Comment by BenjaminLowry** [comment]
@ruben-arts here is my PR that we discussed earlier today, let me know what you think

**Review by baszalmstra — COMMENTED** [review]
I believe conda has more options, @wolfv is that the case? In that case an enum would be nicer.

**Comment by BenjaminLowry** [comment]
@baszalmstra It does look like conda has three options: strict, flexible and disabled https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-channels.html#strict-channel-priority and they default to flexible. Rattler via resolvo is defaulting to strict. I can add an enum but maybe with only two values, strict and disabled for now? And flexible could be added later if needed?

And I see there is a [flag](https://github.com/mamba-org/mamba/blob/80f1e84d49810f0198629da28cd059d457b4c2e6/micromamba/src/server.cpp#L140) for strict priority for libsolv that micromamba uses but I'm a bit ignorant to how this would be translated to this code.

**Comment by baszalmstra** [comment]
Sounds good! Maybe wolf can help with the libsolv side.

**Comment by BenjaminLowry** [comment]
@baszalmstra @wolfv Sorry for the repeated pings, this is just a fairly important change for our workflows.

If we can't resolve why libsolv isn't behaving expectedly with strict channel priority, I can just throw if someone specifies strict when using libsolv? And strict handling could be added later

File: crates/rattler_solve/src/lib.rs
Comment (wolfv) [inline crates/rattler_solve/src/lib.rs:80]: maybe we shoudl use `//` so that these comments do not show up in teh documentation for a feature we don't implement.

File: crates/rattler_solve/src/lib.rs
Comment (wolfv) [inline crates/rattler_solve/src/lib.rs:86]: Shoudl we call this `AllEqual` or `NoPriority/NoPriorities`? Just thinking of something that makes it really clear.

File: crates/rattler_solve/src/libsolv_c/mod.rs
Comment (wolfv) [reply crates/rattler_solve/src/libsolv_c/mod.rs:99]: Yep. Probably we need to set the `priority` on the repository, though? I'll take a look.

**Review by wolfv — COMMENTED** [review]
Just rethinking the naming. Let me take a quick look at libsolv!

File: crates/rattler_solve/src/lib.rs
Comment (BenjaminLowry) [reply crates/rattler_solve/src/lib.rs:80]: Yeah sounds good, I only did that cuz with `//` it formats the comments right against the disabled comment, but likely more confusing for it to be in the documentation

**Comment by wolfv** [comment]
@BenjaminLowry I made a PR against your branch here: https://github.com/BenjaminLowry/rattler/pull/1 It shows how we could set the `priority` for libsolv. 

Libsolv has a `priority` and a `subpriority` integer value. The higher priority should win. We would also need a map or count down the channels and I haven't yet investigated how to do that with the structures at hand.

File: crates/rattler_solve/src/lib.rs
Comment (wolfv) [reply crates/rattler_solve/src/lib.rs:86]: Yeah, totally. I am also OK with keeping the naming. Just sometimes I think we have the opportunity to come up with new & improved terminology :)

**Comment by BenjaminLowry** [comment]
@wolfv Thanks for the PR, I added logic to assign repo priority values where it's determined by the order of the repodatas passed in where the channel with the first repodata is given the highest priority and it descends from there. This is pre-computed so we don't have to do some arbitrary highest priority value and so if you have 5 channels the first will be given 5 and the last will be given 1 and 0 is still available for lowest priority use.

This is now behaving as expected for libsolv with strict and disabled priority, does this implementation seem reasonable? Let me know if I'm not following specific rust-isms as well

**Review by baszalmstra — APPROVED** [review]
LGTM!

**Comment by BenjaminLowry** [comment]
@baszalmstra I'm assuming you also want @wolfv to give a +1? Or is this good to go

**Comment by baszalmstra** [comment]
Yeah since @wolfv has a bunch of open comments I was waiting for his approval as well. 😄 

Wolf if you are reading this, can you merge if you are happy with this?

File: crates/rattler_solve/src/libsolv_c/mod.rs
Comment (wolfv) [inline crates/rattler_solve/src/libsolv_c/mod.rs:167]: These comments are outdated, right?

File: crates/rattler_solve/src/libsolv_c/mod.rs
Comment (wolfv) [inline crates/rattler_solve/src/libsolv_c/mod.rs:165]: Keep or remove?

**Review by wolfv — APPROVED** [review]

**Comment by BenjaminLowry** [comment]
@baszalmstra Thanks for already working on a release! I saw that py-rattler was bumped to 0.4.0 yesterday, will we be able to ship this change in a new release of py-rattler soon?

**Comment by baszalmstra** [comment]
Sure! I can make another release in the next few days.

**Comment by baszalmstra** [comment]
py-rattler 0.5.0 is on pypi which includes your changes.

**Pull request merged**
